src/Wallpaper/Portal.vala

```diff
@@ -0,0 +1,124 @@
+/*
+ * SPDX-FileCopyrigthText: 2024 elementary, Inc. (https://elementary.io)
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ */
+
+[DBus (name = "org.freedesktop.impl.portal.Wallpaper")]
+public class Wallpaper.Portal : Object {
+    private DBusConnection connection;
+
+    public Portal (DBusConnection connection) {
+        this.connection = connection;
+    }
+
+    [DBus (name = "SetWallpaperURI")]
+    public async uint set_wallpaper_uri (
+        ObjectPath handle,
+        string app_id,
+        string parent_window,
+        string uri,
+        HashTable<string, Variant> options
+    ) throws DBusError, IOError {
+        var set_on = "both"; // Possible values are background, lockscreen or both.
+        var show_preview = false;
+
+        if ("set-on" in options && options["set-on"].get_type_string () == "s") {
```

Comment:
This style allows 0 memory allocation and a single lookup on the GLib.HashTable :wink: 
```suggestion
        unowned var set_on = options.get ("set-on");
        if (set_on != null && set_on.is_of_type (GLib.VariantType.STRING)) {
```